### PR TITLE
Add Python Virtualenv Operator Caching

### PR DESCRIPTION
**Pull request by jscheffl**
This PR is a follow-up and split of the PR https://github.com/apache/airflow/pull/33017 to add virtualenv caching to PythonVirtualEnvOperator.

FYI @AutomationDev85 @clellmann @wolfdn

### CODE REVIEW

File: airflow/operators/python.py
Comment (potiuk) [inline airflow/operators/python.py:691]: try/except needs to be handled here. If there is any exception during creation of the venv, it should be entirely deleted.

**Review by potiuk — CHANGES_REQUESTED** [review]
Interesting approach. Glad that you hought about lock and hash @jens-scheffler-bosch . I could think about the case where the venv will be partially created if it fails during cration.

There is also one other case that is difficul to handle. What happens if A virtualenv operator install another dependency after the venv is created. I am not sure if we can do anything about it, but maybe we could store a kind of checksum that would invalidate the venv in case it has been updated after creating it?

**Comment by potiuk** [comment]
> I propose I could add a note to the docs.

Yes. Note in the docs should be sufficient (avoid modifying the venv after it is created or similar ) . checking venv context is not a good idea (not only because of expensiveness of it but also things like .pycaching and potential unintended small but harmless modifications

I also think we should have a way to clean such broken venvs. You know "cache invalidation".  Currently we have no easy/documented way of doing it.

 It does not have to be sophisticated, but one thing that works well is - rather easy - adding a separate unique ID/prefix to such venv so that you can effectively "clean-it" by changing the prefix. This is how it is done in GitHub Actions for example. There you can specify "key" alongside cache definition - and it used to determine if this "the same" or different cache to be used.

In our case you could simply change "path" where the venv is created, but maybe also adding "cache-key" (as sub-folder in the path) might make sense? This way you could store (and sync?) all the cache files by using the same PATH, but then have a key that you could change in case you would like to invalidate this particular venv you use?

File: airflow/operators/python.py
Comment (potiuk) [inline airflow/operators/python.py]: I think we miss a very important issue here. The cache path should also depend on the exact Python version that is used (down to patchlevel). It's obvious for minor versions but we had a problem where new patchlevel of Python could not work with previously prepared venv cache with previous Python version patchlevel.

This was a case once on our CI where we use caching for GitHub actions. The previously prepared venv cache got broken when we updated to a new Python patchlevel version.

Comment (jscheffl) [reply airflow/operators/python.py]: Good idea. Added a `PythonVirtualenvOperator.cache_key` Variable which can be used to create new hashes if needed as admin during runtime (no DAG changes needed).

**Comment by jens-scheffler-bosch** [comment]
Hi @potiuk Thanks for the second round intensive review. Took a moment but now adjusted:
- There is a new Variable `PythonVirtualenvOperator.cache_key` which can optionally be set to influence the hash creation. Also adjusted docs for this
- Added notes for the potential problems if caches venvs are "dirty".
- Added all parameters which influence the venv setup (python version, site packages, index URLs) into the hash.

**Review by potiuk — APPROVED** [review]
All my concerns are addressed. I'd love another pair of eyes to confirm it though.

File: airflow/operators/python.py
Comment (uranusjr) [inline airflow/operators/python.py]:
```suggestion
        venv_cache_path: None | os.PathLike[str] = None,
```

No reason why the user can’t just pass in a Path. (This keeps compatibility with str as well.)

Comment (uranusjr) [reply airflow/operators/python.py]: Might be a good idea to make this a template field too (but for that case we need to move the `Path` cast to `execute` instead)

File: airflow/operators/python.py
Comment (uranusjr) [inline airflow/operators/python.py]: I wonder if this is that worthwhile… Instead of relying on the directory name, we could just dump the values into a JSON file inside the virtual environment directory, and read it out later for cache validation. That could help debugging if caching doesn’t work as expected too.

File: airflow/operators/python.py
Comment (uranusjr) [inline airflow/operators/python.py]: Let’s always use _virtual environment_ (official term in [PEP 405](https://peps.python.org/pep-0405/)) in log messages. The current wording is quite messy, I count _Virtualenv_, _virtualenv_, and _venv_ just in this function.

File: airflow/operators/python.py
Comment (jscheffl) [reply airflow/operators/python.py]: Good points! Applied both changes.

File: airflow/operators/python.py
Comment (jscheffl) [reply airflow/operators/python.py]: I tried to rename all occurrences in the PY file consistently - I hope it is not tooo much now.
Might be (many) other usages of the same term across the docs and other parts of the code as well. If you desire a general "cleanup" of the term, I'd propose to make this in a separate "refactoring" type of PR.

File: airflow/operators/python.py
Comment (jscheffl) [reply airflow/operators/python.py]: There is no cache validation implemented and I assume we should not implement any. If somebody "messes up" the venv then it is anyway hard to track it down.
Requirements file and pip.conf for the install will be kept inside the cached venv is somebody want to debug.

For the moment I'd like to keep the complexity on this level, if there is further demand or issues raised, I think then we can take a look how to harden.
I'd propose if venvs are used, then you should start first with a "normal" one and if you use it often, then the cached version is a performance tuning. But it is not a silver bullet for all sorts of problems. Also there is no pruning and check for sufficient disk space.

The hash is a "simple" approach making the folder from the requirements list idempotent. Hopefully it is only "write once, use many"

Comment (potiuk) [reply airflow/operators/python.py]: > I wonder if this is that worthwhile… Instead of relying on the directory name, we could just dump the values into a JSON file inside the virtual environment directory, and read it out later for cache validation. That could help debugging if caching doesn’t work as expected too.

I agree with @jens-scheffler-bosch "write once, use many" statement.

Validation is not nearly enough. What will you do if the validation fails? Are we going to delete the venv? or recreate it? What happens if we are running in Celery Worker (multiple tasks) and each of them are using "almost" the same environment. For example someone submitted a change to the DAG that will bump a dependency version and this new task is scheduled, but in the same Celery worker we are still running an old versions of the same task from the previous version of the DAG.

Should we delete the venv while another task is running and recreate it? Or should we fail the task? Or maybe wait until the other task completes (how?)

I think really - venvs should be "write once, use many" and immutable after they are created. This is precisely what @jens-scheffler-bosch  addressed well that even if we have two task attempting to create identical venv, we have the file lock that prevents them to do so - and only one will create it and the other will use it. But when it is already created you should not modify existing venv, this is a different story - because you don't want to prevent parallel "read/modify" - you want to prevent parallel "create". You actually want multiple tasks to use the same venv in parallell once created - that's almost the whole point of it.

IMHI calidation is not cutting it because there is no good action that you can attempt once validation fails. We MUST have different venvs if they are different. 

Interestingly - this is for example the very same thing what `pre-commit` does. PRe-commit also calculates hash of the venv - based on the pre-commit definition. If you have two or more tasks with the same "dependencies" and "python version" - they will re-use existing venv and they are even re-used between multiple branches you might work on (see ~/.cache/pre-commit)  and they also get fully recreated with a different hash when anything changes.

**Comment by potiuk** [comment]
cc: @uranusjr - are you ok with the explanations/corrections?

File: airflow/operators/python.py
Comment (uranusjr) [reply airflow/operators/python.py]: By cache validation I meant deciding whether a new task should use an existing environment or create a new one. And instead of trying to encode this as a string in the environment’s path, I feel it’d be easier to just dump the information into JSON, and compare that JSON directly, which is more reliable, likely cheaper, and easier to for post mortem debugging since that file conveniently capture what parameters the environment was created with. Since the implementation already uses a marker file, it can be used to store the JSON as well. This frees us from having to use a pre-determined directory name, which can open up some possible future improvements (I know there are people out there who feel very strongly on the directory name for some reason).

Comment (potiuk) [reply airflow/operators/python.py]: But what would you do if you have two tasks running (old version /new version) with having the same "path".

Imagine you use "rich==2..29" in the PVO, then you have "rich==2.30" change submitted. You have still a task with "rich==2.29" running, but then you also have the next interval scheduled for the same task and it already uses "rich==2.30".

Both tasks want to use the same venv. The first one is running, the second is attempting to run and finds out that rich version changed.

What should we do?

Comment (uranusjr) [reply airflow/operators/python.py]: I do not think it is possible for two versions of a task the exist in the scheduler.

Comment (potiuk) [reply airflow/operators/python.py]: Not the scheduler but the worker.. I think it's perfectly plausible if the two tasks start at different moment (one before the change and one after). Why do you think it's not possible - it's enought that one LocalTaskJob starts (on the same Celery or LocalExecutor worker) just before the "rich" change is done and the other is started "just after".  The first one will run PythonVirtualOperator with "rich==2.29" parameter, and the other with "rich==2.30".

Why do you think its not possible?

Comment (jscheffl) [reply airflow/operators/python.py]: In regards of concurrency... I think "starting almost at the same time" might be a race condition, which we also have seen - I was more having the impression there might be the case that a task runs for an hour and while one is 90% finished a new one might be starting in parallel. Then you need to have the "old" and the "new" venv available on the same worker.

Comment (potiuk) [reply airflow/operators/python.py]: The key part here is that "depedencies, python version, venv path" are all part of the DAG file definition. And that one might (and will) change independently from the tasks being run. There is currently no mechanism in airflow to assure that the task "instance" is using particular version of the DAG Python code. It might change any time and which version of the Python code will be used, depends only on the timing of executing the task on the worker (and how the worker is syncing the DAG code).

When you are backfilling for example, you are all but guaranteed that the same tasks will run for different intervals one after another and if the definition of the venv parametrs changes in the meantime and get synced to the worker, some of the tasks will attempt they PVO runs with "before" and some with "after" the change.

Comment (uranusjr) [reply airflow/operators/python.py]: As Airflow is implemented now backfill tasks actually can only use the latest version of the defined task. There are possible race conditions but honestly those generally would cause worse problems elsewhere than the cached virtual environment because Airflow does not have access to older DAG code.

And OK, if you worry about race condition, it’s of course still fine to put a hash in the virtual environment path. But even with that, having all parameters recorded in the marker file, and using that for comparison, is still better than relying on the hash string IMO.

Comment (potiuk) [reply airflow/operators/python.py]: It's not only race condition. I think it's expected that things will change while running. The backfill case is actually valid. The "latest" you mention is right, but whan you run backfill for multiple dag runs over the span of hours, the `latest` might very easily change mid-flight. I do expect people to bump versions of dependencies frequently in the DAGs.  So it might well be that someone updates the venv while someone else runs the backfill and we have to have a way to handle that case. For now I do not see how we could handle the situation if we keep single venv in this case.

But also there is much more important case. It's ofetn that people will be using  venvs for different tasks even in completely different DAGs. And if we allow for caching and specifyin where to cache the venv, then we have a very high potential of different people using the same paths to cache and causing conflicts. We have currently no mechanism (and we do not plan it) to avoid or solve such conflicts, if we do not use hashes.

Imagine:

```
DAG1:
    PVO(venv_cache_path=/cache/my_venv,requirements="x")

some other team:

DAG2:
    PVO(venv_cache_path=/cache/my_venv,requirements="y")
````

When you get this and keep the venv in the same folder. some of the tasks will start randomly failing (depending on the state of the cache on the worker we are at). If cache is there created by D1 - they D2 will fail, if there is cache created by D2 - D1 tasks will start failing on that worker. The venv will be there but will be a completely different venv. In this case we cannot delete/recreate existing venv when we  see "wrong requirements"  - because the other DAG's tasks might be using it. So the only option we have in case we detect that we want "y" and we see "x" is to fail such task.

I don't think @uranusjr your proposal has a good solution for that case. 

On the other hand `hash` approach is WAY better - it solves it very nicely and it has multiple other advantages that enhance the effect of caching and decrease venv footprint by allowing to share the same venve between seemingly unrelated tasks.

This is precisely why pre-commit hooks have rather small footprint because this is exactly how similar 'hash" approach works there - multiple unrelated pre-commits with same requiremetns share the same venv under-the-hood.

The way how hash approach will work:

* it will not fail randomly - it will **just** work
* even better - it will reuse DAG1s task venv in DAG2 if DAG2 will also have "x" as requirement
* the organisation might even choose to deticate a volume for all caches this way - to reuse and optimise caching and ask ALL their teams to use the same CACHE_FOLDER -this way there is no need to agree additional namspacing and conventions between the teams. The "hash" approach will automaticaly manage creation or reusing the venv when needed in the most optimized way possible.

I really don't see how we can implement it without using hash (but maybe i am missing something?).

Comment (jscheffl) [reply airflow/operators/python.py]: Maybe if this discussion takes longer we need to plan a break-out next week in Toronto to discuss in person to get an agreement :-D

Comment (potiuk) [reply airflow/operators/python.py]: :D

**Comment by jens-scheffler-bosch** [comment]
@uranusjr as discussed during summit added the hash data in the marker as well to detect a hash collision - looking forward for the next level of review :-D

File: airflow/operators/python.py
Comment (uranusjr) [inline airflow/operators/python.py]:
```suggestion
        {"requirements", "index_urls", "venv_cache_path"}.union(PythonOperator.template_fields)
```

Saves a set cast.

File: airflow/operators/python.py
Comment (uranusjr) [inline airflow/operators/python.py]:
```suggestion
        return requirements_hash[:8], hash_text
```

**Comment by uranusjr** [comment]
Some minor comments, no particular objections in general

**Review by uranusjr — APPROVED** [review]

**Comment by potiuk** [comment]
Woohoo! @jens-scheffler-bosch - can you rebase please, just in case ?

**Comment by renpin** [comment]
@jscheffl 

Import fcntl will cause airflow to be unable to run under windows because fcntl seems to be available only on Linux.

I can only go back to 2.7.3 in order to debug dags.

**Comment by potiuk** [comment]
@renpin - Airflow does NOT work on windows. See installation, prerequisites etc.  There are multipele things that won't work on windows - not only that. And all the installation documenation points Windows users to use WSL2 on windows.

There is an open issue for that #10388  and if you want to contribute to it and make airflow compatible with Windows you are free to make it happen. But currently it's not compatible, and it will not work.

**Comment by renpin** [comment]
@potiuk 
Thanks for the answer, but it is easy to debug dag on windows in version 2.7.3. It works well until import "fnctl" in version 2.8.0.
I will try your suggestions or change my development environment to mac.

**Comment by jscheffl** [comment]
> @jscheffl
> 
> Import fcntl will cause airflow to be unable to run under windows because fcntl seems to be available only on Linux.
> 
> I can only go back to 2.7.3 in order to debug dags.

100% agree with @potiuk but was scratching my head. Can you @renpin tell me (1) how you were able to make Airflow running with Airflow 2.7.3 (because I miserably failed to make it working even for simple setups) as well as (2) why do you think this PR is breaking the setup?

**Comment by mullenpaul** [comment]
This PR uses the fcntl package which is simply not supported on Windows. Before 2.8 I could run pytest on my code and now it fails as it can't find the library.

I was also able to get it installed and setup fine locally before this PR. Maybe the answer is for me to raise a pull request making this system agnostic but it isn't great using libraries that are only supported by some operating systems.

**Comment by potiuk** [comment]
> This PR uses the fcntl package which is simply not supported on Windows. Before 2.8 I could run pytest on my code and now it fails as it can't find the library.
> 
> I was also able to get it installed and setup fine locally before this PR. Maybe the answer is for me to raise a pull request making this system agnostic but it isn't great using libraries that are only supported by some operating systems.

Maybe raise PR. As mentioned above - until someone (maybe even you @mullenpaul ) who cares about Windows support will implement https://github.com/apache/airflow/issues/10388 - and get all the tests run and succeed on Windows, there is no way with 30 PRs merged a week that we check every single PR to see if windows compatibility is broken.  So - this is pretty expected that even if accidentally some things will work on Windows sometimes, it's also pretty expected that they will fail equally often and that new PRs will break it. You might ignore the reality, but that's it. This is the reality.

So if you really want to have support for Windows and it is important to you, I suggest. you roll your sleeves up and implement proper support for it including making sure that our CI verifies if things that are supposed to work for windows or not. Airflow is created by > 2800 people and you coudl be one of them. Complaining that this is "not greeat" to use libraries that do not work on Windows when there are no CI checks for it - is just this - idle complainining that brings nothing to the conversation except angering maintainers who work on their free time, taking their time out of their families and possibly paid job so that other people can use the software they help to maintain for absolutely free and without contributing, and yet having demends that THEIR needs are fulfilled.

Yes I suggest rolling sleeves up and implementing - properly with CI tests and everything Windows support. That's the only way to all-but-guarantee it will not be broken in the next 120 or so PRs which we are going to merge during the next month.

Highly recommend it,

**Pull request merged**
